src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProject.cs

```diff
@@ -1470,9 +1497,36 @@ public void ProcessFileChange(string projectSystemFilePath, string workspaceFile
 
                 ClearAndZeroCapacity(_documentsRemovedInBatch);
 
+                // Update project's order of documents.
+                solution = this.UpdateProjectDocumentsOrderForBatch(solution, _project.Id);
+
                 return solution;
             }
 
+            internal Solution UpdateProjectDocumentsOrderForBatch(Solution solution, ProjectId projectId)
+            {
+                // The action of ordering documents was not included in the batch. Simply return the solution.
+                if (_multipleOrderedFilePathsInBatch.Count == 0)
+                {
+                    return solution;
+                }
+
+                var orderedFilePaths = _multipleOrderedFilePathsInBatch.Last();
+
+                ClearAndZeroCapacity(_multipleOrderedFilePathsInBatch);
+
+                // We only care about the latest file paths.
+                return UpdateProjectDocumentsOrder(solution, projectId, orderedFilePaths);
+            }
+
+            private Solution UpdateProjectDocumentsOrder(Solution solution, ProjectId projectId, ImmutableArray<string> filePaths)
+            {
+                var documentIds =
+                    filePaths.Select(x => solution.GetDocumentIdsWithFilePath(x).Single(id => id.ProjectId == projectId));
+
+                return solution.WithProjectDocumentsOrder(projectId, documentIds.ToImmutableList());
```

Comment:
Why are we using `ImmutableList` instead of `immutableArray`?

Reply:
Because in `ProjectState` the internal collection type for documentids is an `ImmutableList`. When we update it, we simply just replace the list with the new one.